File: android/chat-client/src/main/kotlin/com/anapaya/chat/client/ChatError.kt
// Copyright 2026 Anapaya Systems

package com.anapaya.chat.client

/** Everything a call can fail with. */
public sealed class ChatError(
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause) {
    public class Config(detail: String) : ChatError(detail)

    /** The request never produced a response. */
    public class Transport(
        public val kind: TransportFailure,
        detail: String,
        cause: Throwable? = null,
    ) : ChatError("$kind: $detail", cause)

    /** The server refused, and said why. `code` is a string, so a newer code still decodes. */
    public class Api(
        public val status: Int,
        public val code: String,
        public val serverMessage: String,
    ) : ChatError("$code ($status): $serverMessage")

    public class Protocol(detail: String) : ChatError(detail)

    public class NotLoggedIn : ChatError("not logged in")

    public class SessionExpired : ChatError("the session has expired")
}

/** Why a request did not complete, as the transport saw it. */
public enum class TransportFailure {
    Connectivity,

    Resolution,

    Connect,

    Tls,

    StreamReset,

    Protocol,

    BodyTooLarge,

    Timeout,

    Closed,
}

File: ios/ChatClient/Sources/ChatClient/ChatError.swift
// Copyright 2026 Anapaya Systems

import Foundation

/// Why a transport could not carry a request.
public enum TransportFailure: Sendable, Equatable {
    case connectivity
    case resolution
    case connect
    case tls
    case streamReset
    case timeout
    case bodyTooLarge
    case protocolError
}

/// Everything a call can fail with.
public enum ChatError: Error, Sendable, Equatable {
    case config(String)
    /// The request never produced a response.
    case transport(TransportFailure, String)
    case protocolError(String)
    case api(status: Int, code: String, message: String)
    case notLoggedIn
    case sessionExpired
}

extension ChatError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .config(let detail):
            return detail
        case .transport(_, let detail):
            return detail
        case .protocolError(let detail):
            return detail
        case .api(let status, let code, let message):
            return "\(code) (\(status)): \(message)"
        case .notLoggedIn:
            return "not logged in"
        case .sessionExpired:
            return "the session has ended"
        }
    }
}
